### PR TITLE
[state_sync] Cleanup forced state snapshot created by resharding

### PR DESCRIPTION
**Pull request by shreyan-gupta**
This PR handles the task of cleaning up snapshots that were produced after resharding.

For context, in this PR https://github.com/near/nearcore/pull/9681 we are overriding the default state_snapshot enabled config to force the production of a snapshot for resharding. We need to worry about later cleaning up/deleting this snapshot.

The simplest implementation calls for us to clear/delete the snapshot on the `make_state_snapshot` call. Initially we were returning early in case the state_snapshot config was disabled (except when forced). The new behavior is to always delete old snapshot.

### CODE REVIEW

File: core/store/src/trie/state_snapshot.rs [hunk 1]
```diff
@@ -160,46 +160,55 @@ impl ShardTries {
...
 
+        // Drop Store before deleting the underlying data.
+        // We delete the snapshot whether the config is enabled or not. This is to delete any potential
+        // snapshots that may have been produced during resharding, even if the config is disabled.
+        *state_snapshot_lock = None;
+        self.delete_current_snapshot();
+        metrics::HAS_STATE_SNAPSHOT.set(0);
+
+        // override enabled config if force_snapshot is true
+        if !self.state_snapshot_config().enabled && !force_snapshot {
```
Comment (shreyan-gupta) [inline core/store/src/trie/state_snapshot.rs:195]: main change: reorder to first call `self.delete_current_snapshot();` and then proceed with `if !self.state_snapshot_config().enabled && !force_snapshot`

File: integration-tests/src/tests/client/resharding.rs [hunk 2]
```diff
@@ -605,6 +606,28 @@ impl TestReshardingEnv {
             }
         }
     }
+
+    fn check_snapshot(&mut self, state_snapshot_enabled: bool) {
```
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/resharding.rs:610]: In the previous PR https://github.com/near/nearcore/pull/9681, we added a test to check if resharding happens when state snapshot config is enabled or disabled. In this PR we extend the tests to check whether the snapshot is deleted or not.

**Comment by akhi3030** [comment]
Removing myself in favour of wacban.

File: integration-tests/src/tests/client/resharding.rs (hunk 2 above)
Comment (walnut-the-cat) [reply integration-tests/src/tests/client/resharding.rs:611]: what does 'check_snapshot' mean? (check 'What' of snapshot?) function name doesn't sound clear enough.

Comment (wacban) [reply integration-tests/src/tests/client/resharding.rs:611]: I'm fine with the function name as it's similar to others but please add a comment :)

**Review by wacban — APPROVED** [review]
LGTM

**Pull request closed without merging**